File: ai-agent/app/utils/gemini_client.py
import json
import re
import time
import asyncio
from app.utils.config import settings

try:
    import google.generativeai as genai
    GEMINI_AVAILABLE = True
except ImportError:
    GEMINI_AVAILABLE = False

def _strip_markdown_json(text: str) -> str:
    """Removes markdown backticks and formatting from JSON string blocks."""
    text = text.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(json)?\n", "", text, flags=re.IGNORECASE)
        text = re.sub(r"\n```$", "", text)
    return text.strip()

def generate(prompt: str, system: str) -> str:
    """
    Calls Gemini API and returns the response text.
    Synchronous version for compatibility.
    """
    if not GEMINI_AVAILABLE:
        raise RuntimeError("google-generativeai not installed. Install with: pip install google-generativeai")
    
    if not settings.GEMINI_API_KEY:
        raise ValueError("GEMINI_API_KEY not set in environment")
    
    genai.configure(api_key=settings.GEMINI_API_KEY)
    model = genai.GenerativeModel("gemini-2.5-flash")
    full_prompt = f"{system}\n\n{prompt}"
    
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = model.generate_content(full_prompt)
            return response.text
        except Exception as e:
            error_msg = str(e)
            if "429" in error_msg or "Quota exceeded" in error_msg:
                if attempt < max_retries - 1:
                    print(f"Gemini API Rate Limit hit. Waiting 35 seconds before retry {attempt + 1}/{max_retries}...")
                    time.sleep(36.0)
                    continue
            raise ValueError(f"Gemini API error: {error_msg}")

async def generate_async(prompt: str, system: str) -> str:
    """
    Async wrapper for Gemini API call.
    """
    if not GEMINI_AVAILABLE:
        raise RuntimeError("google-generativeai not installed. Install with: pip install google-generativeai")
    
    if not settings.GEMINI_API_KEY:
        raise ValueError("GEMINI_API_KEY not set in environment")
    
    genai.configure(api_key=settings.GEMINI_API_KEY)
    model = genai.GenerativeModel("gemini-2.5-flash")
    full_prompt = f"{system}\n\n{prompt}"
    
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = await model.generate_content_async(full_prompt)
            return response.text
        except Exception as e:
            error_msg = str(e)
            if "429" in error_msg or "Quota exceeded" in error_msg:
                if attempt < max_retries - 1:
                    print(f"Gemini API Rate Limit hit. Waiting 35 seconds before retry {attempt + 1}/{max_retries}...")
                    await asyncio.sleep(36.0)
                    continue
            raise ValueError(f"Gemini API error: {error_msg}")

def generate_json(prompt: str, system: str) -> dict:
    """
    Calls Gemini API and parses response as JSON.
    Raises ValueError if parsing fails.
    """
    system_with_json_directive = f"{system}\n\nYou MUST respond with valid JSON only."
    
    response_text = generate(prompt, system_with_json_directive)
    cleaned_text = _strip_markdown_json(response_text)
    
    try:
        parsed = json.loads(cleaned_text)
        return parsed
    except json.JSONDecodeError as e:
        raise ValueError(f"Failed to parse Gemini JSON response: {str(e)}\nResponse was: {cleaned_text}")

async def generate_json_async(prompt: str, system: str) -> dict:
    """
    Async version: Calls Gemini API and parses response as JSON.
    Raises ValueError if parsing fails.
    """
    system_with_json_directive = f"{system}\n\nYou MUST respond with valid JSON only."

    response_text = await generate_async(prompt, system_with_json_directive)
    cleaned_text = _strip_markdown_json(response_text)

    try:
        parsed = json.loads(cleaned_text)
        return parsed
    except json.JSONDecodeError as e:
        raise ValueError(f"Failed to parse Gemini JSON response: {str(e)}\nResponse was: {cleaned_text}")

def validate_gemini_connection():
    try:
        result = generate("Hello", "You are a test assistant.")
        return {
            "status": "success",
            "message": "Gemini connection successful",
            "response_preview": result[:100]
        }
    except Exception as e:
        return {
            "status": "error",
            "message": str(e)
        }

File: escrow-service/src/utils/gemini.client.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import axios from 'axios';
import { logger } from './logger';

function stripMarkdownJson(text: string): string {
  text = text.trim();
  if (text.startsWith('```')) {
    text = text.replace(/^```(json)?\n/i, '').replace(/\n```$/, '');
  }
  return text.trim();
}

export async function generateJson(prompt: string, system: string): Promise<any> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error('GEMINI_API_KEY not set in environment');
  }

  const genai = new GoogleGenerativeAI(apiKey);
  const model = genai.getGenerativeModel({ model: 'gemini-2.5-flash' });

  const systemWithJsonDirective = `${system}\n\nYou MUST respond with valid JSON only.`;
  const fullPrompt = `${systemWithJsonDirective}\n\n${prompt}`;

  const maxRetries = 3;
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      const result = await model.generateContent(fullPrompt);
      const responseText = result.response.text();
      const cleaned = stripMarkdownJson(responseText);

      try {
        return JSON.parse(cleaned);
      } catch (parseError) {
        throw new Error(`Failed to parse Gemini JSON response: ${(parseError as Error).message}\nResponse was: ${cleaned}`);
      }
    } catch (error) {
      const msg = (error as Error).message || '';
      const isRateLimit = msg.includes('429') || msg.toLowerCase().includes('quota exceeded') || msg.toLowerCase().includes('resource exhausted');

      if (isRateLimit && attempt < maxRetries - 1) {
        logger.info(`Gemini rate limit hit, waiting 36s before retry ${attempt + 1}/${maxRetries}`);
        await new Promise(resolve => setTimeout(resolve, 36000));
        continue;
      }
      throw error;
    }
  }

  throw new Error('Gemini API failed after max retries');
}

export async function analyzeDesignImage(
  imageUrl: string,
  requirements: string[]
): Promise<{ pass: boolean; score: number; evidence: string }> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) throw new Error('GEMINI_API_KEY not set');

  let imageData: string;
  let mimeType: string;

  try {
    const response = await axios.get(imageUrl, { responseType: 'arraybuffer', timeout: 15000 });
    imageData = Buffer.from(response.data as ArrayBuffer).toString('base64');
    mimeType = (response.headers['content-type'] as string) || 'image/png';
  } catch (err) {
    throw new Error(`Failed to fetch design image: ${(err as Error).message}`);
  }

  const genai = new GoogleGenerativeAI(apiKey);
  const model = genai.getGenerativeModel({ model: 'gemini-2.5-flash' });

  const prompt = `You are a professional UI/UX design reviewer. Analyze this design image against the following requirements:
${requirements.map((r, i) => `${i + 1}. ${r}`).join('\n')}

Respond ONLY with valid JSON (no markdown, no explanation):
{"pass": boolean, "score": number_0_to_100, "evidence": "brief_explanation_of_what_matches_or_is_missing"}`;

  const result = await model.generateContent([
    { text: prompt },
    { inlineData: { mimeType, data: imageData } }
  ]);

  const text = stripMarkdownJson(result.response.text());
  try {
    return JSON.parse(text);
  } catch {
    throw new Error(`Failed to parse Gemini vision response: ${text}`);
  }
}
